Source/NSStoryboard.m

```diff
@@ -0,0 +1,975 @@
+        }
+      if ([coder containsValueForKey: @"NSSender"])
+        {
+          [self setSender: [coder decodeObjectForKey: @"NSSender"]];
+        }
+      if ([coder containsValueForKey: @"NSIdentifier"])
+        {
+          [self setIdentifier: [coder decodeObjectForKey: @"NSIdentifier"]];
+        }
+      if ([coder containsValueForKey: @"NSKind"])
+        {
+          [self setKind: [coder decodeObjectForKey: @"NSKind"]];
+        }
+    }
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  // this is never encoded directly...
+}
+@end
+
+
+@interface NSControllerPlaceholder : NSObject <NSCoding, NSCopying> // , NSSeguePerforming>
+{
+  NSString *_storyboardName;
+  NSStoryboard *_storyboard;
+}
+
+- (NSString *) storyboardName;
+- (void) setStoryboardName: (NSString *)name;
+- (id) instantiate;
+
+@end
+
+@implementation NSControllerPlaceholder
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      _storyboardName = nil;
+    }
+  return self;
+}
+
+- (NSString *) storyboardName
+{
+  return _storyboardName;
+}
+
+- (void) setStoryboardName: (NSString *)name
+{
+  ASSIGNCOPY(_storyboardName, name);
+}
+
+- (id) copyWithZone: (NSZone *)z
+{
+  NSControllerPlaceholder *c = [[NSControllerPlaceholder allocWithZone: z] init];
+  [c setStoryboardName: _storyboardName];
+  return c;
+}
+
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  self = [super init];
+  if ([coder allowsKeyedCoding])
+    {
+      if ([coder containsValueForKey: @"NSStoryboardName"])
+        {
+          [self setStoryboardName: [coder decodeObjectForKey: @"NSStoryboardName"]];
+        }
+    }
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  // this is never encoded directly...
+}
+
+- (id) instantiate
+{
+  _storyboard = [NSStoryboard storyboardWithName: _storyboardName
+                                          bundle: [NSBundle mainBundle]];
+  return [_storyboard instantiateInitialController];
+}
+
+@end
+
+@implementation NSStoryboard
+
+- (NSXMLElement *) _createCustomObjectWithId: (NSString *)ident
+                                   userLabel: (NSString *)userLabel
+                                 customClass: (NSString *)className
+{
+  NSXMLElement *customObject =
+    [[NSXMLElement alloc] initWithName: @"customObject"];
+  NSXMLNode *idValue =
+    [NSXMLNode attributeWithName: @"id"
+                     stringValue: ident];
+  NSXMLNode *usrLabel =
+    [NSXMLNode attributeWithName: @"userLabel"
+                     stringValue: userLabel];
+  NSXMLNode *customCls =
+    [NSXMLNode attributeWithName: @"customClass"
+                     stringValue: className];
+  
+  [customObject addAttribute: idValue];
+  [customObject addAttribute: usrLabel];
+  [customObject addAttribute: customCls];
+
+  AUTORELEASE(customObject);
+  
+  return customObject;
+}
+
+- (void) _processStoryboard: (NSXMLDocument *)storyboardXml
```

Comment:
Wouldn't this whole method be better expressed as an XSLT?

Reply:
Most likely, but it's done now and it is working perfectly.  It would be interesting to write a transform for it, but I will save that for the refactor.

Reply:
Deferred for refactoring.  I think doing this via XSLT might make it much more compact.  Doing it in ObjC code helped me understand everything which needs to be done.  What would concern me is making XSLT a hard dependency to support this.

Reply:
Done

Reply:
I really don't think we are ever going to re-write this in XSLT.